optimism/src/preimage_oracle.rs

```diff
@@ -24,6 +24,57 @@ pub struct ReadWrite<R, W> {
 
 pub struct RW(pub ReadWrite<PipeReader, PipeWriter>);
 
+// Here, we implement `os_pipe::pipe` in a way that allows us to pass flags. In particular, we
+// don't pass the `CLOEXEC` flag, because we want these pipes to survive an exec, and we set
+// `DIRECT` to handle writes as single atomic operations (up to splitting at the buffer size).
+// This fixes the IPC hangs. This is bad, but the hang is worse.
+
+#[cfg(not(any(target_os = "ios", target_os = "macos", target_os = "haiku", windows)))]
+fn create_pipe() -> std::io::Result<(PipeReader, PipeWriter)> {
+    let mut fds: [libc::c_int; 2] = [0; 2];
+    let res = unsafe { libc::pipe2(fds.as_mut_ptr(), libc::O_DIRECT) };
+    if res != 0 {
+        return Err(std::io::Error::last_os_error());
+    }
+    unsafe {
+        use std::os::fd::FromRawFd;
+        Ok((
+            PipeReader::from_raw_fd(fds[0]),
+            PipeWriter::from_raw_fd(fds[1]),
+        ))
+    }
+}
+
+#[cfg(any(target_os = "ios", target_os = "macos", target_os = "haiku"))]
+pub fn create_pipe() -> std::io::Result<(PipeReader, PipeWriter)> {
+    let mut fds: [libc::c_int; 2] = [0; 2];
+    let res = unsafe { libc::pipe(fds.as_mut_ptr()) };
+    if res != 0 {
+        return Err(std::io::Error::last_os_error());
+    }
+    // It appears that Mac and friends don't have DIRECT. Oh well. Don't use a Mac.
+    let res = unsafe { libc::fcntl(fds[0], libc::F_SETFD, 0) };
+    if res != 0 {
+        return Err(std::io::Error::last_os_error());
+    }
+    let res = unsafe { libc::fcntl(fds[1], libc::F_SETFD, 0) };
+    if res != 0 {
+        return Err(std::io::Error::last_os_error());
+    }
+    unsafe {
+        use std::os::fd::FromRawFd;
+        Ok((
+            PipeReader::from_raw_fd(fds[0]),
+            PipeWriter::from_raw_fd(fds[1]),
+        ))
+    }
+}
+
+#[cfg(windows)]
+pub fn create_pipe() -> std::io::Result<(PipeReader, PipeWriter)> {
+    os_pipe::pipe()
+}
+
```

Comment:
Do we actually hope to run any of this on Windows? If not this would remove the dependency on `os_pipe`

Reply:
I'm sure somebody will try at some point :)

Reply:
Right. You could craft a carefully worded comment for those uses as for the Mac users 😈